server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -252,6 +240,33 @@ private static long newTargetIORateBytesPerSec(
         return newTargetIORateBytesPerSec;
     }
 
+    static class AtomicIORate extends AtomicLong {
+
+        public AtomicIORate(long initialIORate) {
+            super(initialIORate);
+        }
+
+        // Exactly like {@link AtomicLong#updateAndGet} but calls the consumer rather than return the new (updated) value.
+        // The consumer receives both the previous and the updated values (which can be equal).
+        void update(LongUnaryOperator updateFunction, UpdateConsumer updateConsumer) {
+            long prev = get(), next = 0L;
+            for (boolean haveNext = false;;) {
+                if (!haveNext)
+                    next = updateFunction.applyAsLong(prev);
+                if (weakCompareAndSetVolatile(prev, next)) {
```

Comment:
Should we only do this if prev != next, otherwise not update anything? Trying to avoid unnecessary contention on the field.

Reply:
Maybe... that's how I thought of it in the previous implementation.
But 1) it is an extra implementation delta from the `AtomicLong#updateAndGet` implementation and 2) this implementation difference, i.e. with the state where the `new == prev` but `compareAndSet(prev, new)` fails is not easy to explain (especially when `new` is not a mathematical function of `prev`)

Reply:
To explain why it is difficult to explain, even if `new` equals `prev`, if `prev` is not still the current value, the update should be retried (as if `new` was not equal to `prev`, because `new` equaling `prev` at some point in time is not meaningful as starting from the same `prev`, at some other point in time, would result in a different `new`).

Reply:
In other words, struggling to document/explain the above away is not worth the contention savings in this case, I believe.

Reply:
Sounds good, thanks for the elaboration.